File: Problems/Algorithms/542. 01 Matrix/01_matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        n, m = len(mat), len(mat[0])
        results = [[n*m if mat[i][j] == 1 else 0 for j in range(m)] for i in range(n)]

        queue = []
        for i in range(n):
            for j in range(m):
                if mat[i][j] == 0:
                    queue.append([i, j])

        neighbors = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        while len(queue) > 0:
            r, c = queue.pop(0)

            for neighbor in neighbors:
                x, y = r + neighbor[0], c + neighbor[1]
                if self.is_valid(n, m, x, y):
                    if results[x][y] > results[r][c] + 1:
                        results[x][y] = results[r][c] + 1
                        queue.append([x, y])

        return results

    @staticmethod
    def is_valid(n: int, m: int, r: int, c: int) -> bool:
        return r >= 0 and c >= 0 and r < n and c < m

File: Problems/Algorithms/542. 01 Matrix/01_matrix.go
func updateMatrix(mat [][]int) [][]int {
    n, m := len(mat), len(mat[0])
    results := make([][]int, n)
    for row := range results {
        results[row] = make([]int, m)
    }

    queue := make([][2]int, 0)
    for i := 0; i < n; i++ {
        for j:= 0; j < m; j++ {
            if mat[i][j] == 0 {
                queue = append(queue, [2]int{i, j})
            } else {
                results[i][j] = n * m
            }
        }
    }

    neighbors := [4][2]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}
    for len(queue) > 0 {
        r, c := queue[0][0], queue[0][1]
        queue = queue[1:]

        for _, neighbor := range neighbors {
            x, y := r + neighbor[0], c + neighbor[1]
            if isValid(n, m, x, y) {
                if results[x][y] > results[r][c] + 1 {
                    results[x][y] = results[r][c] + 1
                    queue = append(queue, [2]int{x, y})
                }
            }
        }
    }

    return results
}

func isValid(n int, m int, r int, c int) bool {
    return r >= 0 && c >= 0 && r < n && c < m
}
